clippy_lints/src/casts/needless_type_cast.rs

```diff
@@ -0,0 +1,199 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet;
+use clippy_utils::visitors::for_each_expr_without_closures;
+use core::ops::ControlFlow;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_errors::Applicability;
+use rustc_hir::def::Res;
+use rustc_hir::{Block, Body, ExprKind, HirId, LetStmt, PatKind, StmtKind};
+use rustc_lint::LateContext;
+use rustc_middle::ty::Ty;
+use rustc_span::Span;
+
+use super::NEEDLESS_TYPE_CAST;
+
+struct BindingInfo<'a> {
+    source_ty: Ty<'a>,
+    ty_span: Option<Span>,
+    pat_span: Span,
+}
+
+struct UsageInfo<'a> {
+    is_cast: bool,
+    cast_to: Option<Ty<'a>>,
+}
+
+pub(super) fn check<'a>(cx: &LateContext<'a>, body: &Body<'a>) {
+    let mut bindings: FxHashMap<HirId, BindingInfo<'a>> = FxHashMap::default();
+
+    collect_bindings_from_block(cx, body.value, &mut bindings);
+
+    for_each_expr_without_closures(body.value, |expr| {
+        if let ExprKind::Let(let_expr) = expr.kind {
+            collect_binding_from_let(cx, let_expr, &mut bindings);
+        }
+        ControlFlow::<()>::Continue(())
+    });
+
+    #[allow(rustc::potential_query_instability)]
+    let mut binding_vec: Vec<_> = bindings.into_iter().collect();
+    binding_vec.sort_by_key(|(_, info)| info.pat_span.lo());
+
+    for (hir_id, binding_info) in binding_vec {
+        check_binding_usages(cx, body, hir_id, &binding_info);
+    }
+}
+
+fn collect_bindings_from_block<'a>(
+    cx: &LateContext<'a>,
+    expr: &rustc_hir::Expr<'a>,
+    bindings: &mut FxHashMap<HirId, BindingInfo<'a>>,
+) {
+    if let ExprKind::Block(block, _) = expr.kind {
+        collect_bindings_from_block_inner(cx, block, bindings);
+    }
+}
+
+fn collect_bindings_from_block_inner<'a>(
+    cx: &LateContext<'a>,
+    block: &Block<'a>,
+    bindings: &mut FxHashMap<HirId, BindingInfo<'a>>,
+) {
+    for stmt in block.stmts {
+        if let StmtKind::Let(let_stmt) = stmt.kind {
+            collect_binding_from_local(cx, let_stmt, bindings);
+        }
+    }
+}
+
+fn collect_binding_from_let<'a>(
+    cx: &LateContext<'a>,
+    let_expr: &rustc_hir::LetExpr<'a>,
+    bindings: &mut FxHashMap<HirId, BindingInfo<'a>>,
+) {
+    if let_expr.ty.is_none() {
```

Comment:
This lint should have a `form_expansion` check to avoid linting code in macros.

Yes, it's possible that there may be wins, but often macros have been written a certain way for a reason, and this might reduce the noise the lint makes.